### PR TITLE
[wip][update] fs: Background commit in complete multipart upload

### LINKED ISSUE #2334: Don't concatenate multi-part upload during completion
The FS version concatenates all the parts of a multi-part upload when the multi-upload is completed. Multi-part uploads are often used for huge files, so this takes a lot of time and also spikes the IOPS for a while. In some situations (like ours), it would be much better to use the XL method to keep the parts separate and combine them during download.

Of course, the parts cannot be accessed the regular way (via the filesystem) anymore. But most systems that use S3 only use S3 to access them (you can't do otherwise in AWS), so this wouldn't be a big problem.

Concatenating files might be the sensible default option, but when performance and I/O is critical, then keeping the files separate might be the better option. Now, we reverted using the XL method, solely for this feature, but this costs us a lot of additional disk space.

### PR DESCRIPTION
**Pull request by vadmeste**
Some shortcomings with this patch:
*) the object will be not seen just after upload until the background commit will terminate

Fixes #2334

### CODE REVIEW

**Comment by vadmeste** [comment]
Also, one test unit case is removed, which was testing is an upload id is invalid after a successful multipart complete (since old upload ids are removed), we should not check this anymore since removing the upload id is put in a goroutine

**Comment by krishnasrinivas** [comment]
> we should not check this anymore since removing the upload id is put in a goroutine

you can Sleep() before checking because the test case is valid.

**Comment by vadmeste** [comment]
> you can Sleep() before checking because the test case is valid.

yes we can, but it is just how much time it will sleep.. 1 seconds, 10 seconds? on travis it could be 1 min, that's why I am suggested to remove this one even if it is an important one, anyway let's discuss the overall feature before we talk about this particular point.

File: fs-v1-multipart.go [hunk 1]
```diff
@@ -542,91 +537,132 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
...
 		multipartPartFile := path.Join(mpartMetaPrefix, bucket, object, uploadID, partSuffix)
-		offset := int64(0)
-		totalLeft := fsMeta.Parts[partIdx].Size
-		for totalLeft > 0 {
-			curLeft := int64(readSizeV1)
-			if totalLeft < readSizeV1 {
-				curLeft = totalLeft
+		_, err = fs.storage.StatFile(minioMetaBucket, multipartPartFile)
+		if err != nil {
+			return "", err
```
Comment (harshavardhana) [inline fs-v1-multipart.go]: You need to see a different error here `InvalidPart{}`

File: fs-v1-multipart.go [hunk 2]
```diff
@@ -542,91 +537,132 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
...
+		}
+
+	}
+
+	// Background commit
+	go func() {
```
Comment (harshavardhana) [inline fs-v1-multipart.go]: Keep this in a separate function, take inputs for all the parts.

File: fs-v1-multipart.go [hunk 3]
```diff
@@ -542,91 +537,132 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
...
+				//FIXME: alert if needed
 			}
-			var n int64
-			n, err = fs.storage.ReadFile(minioMetaBucket, multipartPartFile, offset, buf[:curLeft])
-			if n > 0 {
-				if err = fs.storage.AppendFile(minioMetaBucket, tempObj, buf[:n]); err != nil {
-					return "", toObjectErr(err, minioMetaBucket, tempObj)
+		}()
+		// Hold lock in the uploaded object until background commiting terminates
+		nsMutex.Lock(bucket, object)
```
Comment (harshavardhana) [inline fs-v1-multipart.go]: No need to hold lock until 'rename'.

**Comment by harshavardhana** [comment]
Can you rebase this patch @vadmeste ?

File: fs-v1-multipart.go [hunk 4]
```diff
@@ -542,6 +539,48 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
...
+		}
+
+		totalSize += fsMeta.Parts[partIdx].Size
+	}
+
+	// Check if we have enough disk space to make the commit. The parts size + the target file (totalSize*2)
+	// should respect the minimum free space requirement.
+	storageInfo := fs.StorageInfo()
+	if storageInfo.Free+totalSize*2 < storageInfo.Total*(1-fsMinSpacePercent) {
+		return "", errDiskFull
```
Comment (harshavardhana) [inline fs-v1-multipart.go]:
```
toObjectErr(errDiskFull, bucket, object) 
```

**Comment by harshavardhana** [comment]
@vadmeste is this still a WIP?

**Comment by vadmeste** [comment]
@harshavardhana, no, will do your notice and sync again

**Comment by vadmeste** [comment]
Do not merge yet, one problem with this patch is that it is modifying original file (os.Truncate) during complete multipart

**Comment by harshavardhana** [comment]
@vadmeste  we decided to do a separate backend for this by making XL itself support Single disk and multi disk.

Single disk means -- no erasure but all other things are available (saving metadata and bit-rot)
Multiple disks means -- all things same as is.

**Comment by vadmeste** [comment]
Brilliant simplification!

**Comment by harshavardhana** [comment]
This fix is not needed anymore.. @vadmeste you can close.

**Pull request closed without merging**